packages/core/src/Popout/Popout.tsx

```diff
@@ -0,0 +1,134 @@
+import React, { RefObject, useCallback, useEffect, useLayoutEffect, useRef, useState } from 'react'
+import styled from 'styled-components'
+import { Box } from '..'
+
+export interface IPopoutProps {
+  trigger: JSX.Element
+  content?: JSX.Element | undefined
+  triggerRef?: RefObject<HTMLElement>
+  onOpen?: () => void
+  onClose?: () => void
+  closeOnTriggerRefClick?: boolean
+}
+
+const TriggerWrapper = styled(Box)`
+  position: relative;
+  width: fit-content;
+`
+
+const OnClickHandler = styled(Box)`
+  position: absolute;
+  top: 0;
+  width: 100%;
+  height: 100%;
+  cursor: pointer;
+`
+
+const Shadow = styled(Box)`
+  position: fixed;
+  top: 0;
+  left: 0;
+  cursor: pointer;
+  width: 100vw;
+  height: 100vh;
+  z-index: 1;
+  transition: opacity 250ms ease-in-out;
+`
+
+const PopoutModal = styled(Box)`
+  position: absolute;
+  z-index: 2;
+  background-color: white;
+  transition: padding 200ms ease-in-out, margin 200ms ease-in-out, border-radius 200ms ease-in-out;
+`
+
+export const Popout = (props: IPopoutProps) => {
+  const { trigger, content, onOpen, onClose, triggerRef, closeOnTriggerRefClick } = props
+  const [isOpen, setIsOpen] = useState(false)
+  const [height, setHeight] = useState(0)
+  const [padding, setPadding] = useState(0)
+  const [opacity, setOpacity] = useState(0)
+  const baseInputRef = useRef<HTMLDivElement>(null)
+  const modalRef = useRef<HTMLDivElement>(null)
+
+  const handleOpen = useCallback(() => {
+    setIsOpen(true)
+    setTimeout(() => {
+      setPadding(16)
+      setOpacity(0.5)
+      /* istanbul ignore next */
+      triggerRef?.current?.focus()
+      /* istanbul ignore next */
+      onOpen?.()
+    })
```

Comment:
why is there a `setTimeout` here? this implies there is an issue with state

Reply:
You might want to try putting it on a `useEffect` with `triggerRef` and `isOpen` in the dependency array and try removing the `setTimeout` and see if that still works. You can then add an if statement for `isOpen`.

Reply:
I am a bit embarassed that I didn't realize this should have been a useEffect the whole time 🤦‍♂️ it seems so obvious now. Implemented the suggestions @ryan-hunt-priceline recommended and it works perfectly, and is much more readable I feel. Thank you both for being so thorough and helpful with your feedback!